### DISCUSSION TITLE
Confusion around auth_token/refresh_token with SSR

### DISCUSSION THREAD

**Discussion by genu** [directus/directus#15079] [Archive]
I've been trying to figure this out for some time now, but I keep hitting road blocks due to conflicting information between docs, discussions, source code

Has anyone gotten Nuxt to work with directus?

Some of my biggest pain points:

- Directus is intialized with the `MemoryStorage` on SSR and `LocalStorage` on client
     - Login works on client and auth token is stored in local storage, but every page refresh will then throw auth errors because SSR would not have the token
     - To Over come this, I tried to sync `LocalStorage` and `MemoryStorage` in nuxt using `useCookie`
     
     So now, I have a middleware in nuxt that looks like this:
     
     ```ts
      if (process.server) {
    addRouteMiddleware(
      'sync-auth-tokens',
      () => {
        const auth_token = useCookie('auth_token').value
        const auth_refresh_token = useCookie('auth_refresh_token').value
        const auth_expires = useCookie<number>('auth_expires').value

        if (auth_token && auth_refresh_token && auth_expires) {
          directus.storage.auth_token = useCookie('auth_token').value
          directus.storage.auth_refresh_token = useCookie('auth_refresh_token').value
          directus.storage.auth_expires = useCookie<number>('auth_expires').value
        }
      },
      { global: true }
    )
  }
  ```
  
 - But now I have the token information stored in two places.
 
This solves the problem of page refreshes, but ads some duplication and overhead.

 - **Another Pain point:** When a token expires, SSR doesn't use the refresh token.
    - After some time, the token expires.
    - On Client, the token is refreshed, but on SSR, this doesn't happen.
    - I need to find a way to create a hook for when the token is refreshed to also update it in the cookie so that I can sync it with the SSR version.
    
    Is there a way to subscribe to a token change, or watch it for changes so that I can sync it with SSR?

**Comment by azrikahar** [comment]
> Nuxt to work with directus

Based on #15068, I assume `directus` here means specifically with Directus SDK. In that case, you can refer to this helpful plugin by @bryantgillespie here which integrates the SDK, and seems to use a custom CookieStorage class: https://github.com/bryantgillespie/nuxt3-directus-starter/blob/main/plugins/directus.ts

Perhaps you can give it a try to see whether it addresses the pain points you are having?
